src/common/shared_cache.hpp

```diff
@@ -39,6 +40,65 @@ class SharedLRU {
   using VPtr = std::shared_ptr<V>;
   using WeakVPtr = std::weak_ptr<V>;
 #endif
+
+  class Cleanup {
+    SharedLRU<K, V>* cache;
+    K key;
+    static_assert(std::is_trivially_destructible_v<decltype(cache)>);
+
+  public:
+    Cleanup(SharedLRU<K, V>* const cache, K key)
+      : cache(cache),
+	key(key) {
+    }
+    void operator()(V* const ptr) {
+      cache->remove(key, ptr);
+      delete ptr;
+    }
+  };
+
+  // https://stackoverflow.com/questions/32113594/weak-ptr-make-shared-and-memory-deallocation
+  template <class T>
+  class VCleaningAlloc {
+  public:
+    SharedLRU<K, V>* cache;
+    K key;
+
+    typedef T value_type;
+
+    VCleaningAlloc(SharedLRU<K, V>* const cache, K key)
+      : cache(cache),
+	key(key) {
+    }
+
+    template<typename U>
+    VCleaningAlloc(const VCleaningAlloc<U>& other)
+      : cache(other.cache),
+	key(other.key) {
+    }
+
+    T* allocate(const std::size_t n) {
+      if (auto ptr = std::malloc(n * sizeof(T))) {
+        return static_cast<T*>(ptr);
+      }
+      throw std::bad_alloc();
+    }
+
+    void destroy(T* ptr) {
+      cache->remove(key, ptr);
```

Comment:
but i think the customised deleter of shared_ptr already addresses the issue where weak_ptr outlives the shared_ptr. as we always delete the weak_ptr once all shared_ptr instances are eliminated. the customized allocator fixes this issue in a different way though.

why do you think it's a better solution?